### PR TITLE
Rwjs/anifixd21

### PR DESCRIPTION
**Pull request by robsimmonds**
Fix for the queuing metrix code.
Added an additional task to avoid possible race condition when starting an animation if there is an existing animation running.

### CODE REVIEW

**Comment by veggiesaurus** [comment]
@robsimmonds can you merge `dev` into your branch so that the ICD tests pass

File: src/Concurrency.h [hunk 1]
```diff
@@ -83,21 +82,26 @@ class queuing_rw_mutex {
         if ((_reader_count > 0) || (_writer_count > 1)) {
             queue_writer(lock);
         }
+        short rc2 = _reader_count;
```
Comment (veggiesaurus) [inline src/Concurrency.h]: What's the point of these two lines? (and the ones below). `rc` and `wc` are never used.

File: src/Session.h [hunk 2]
```diff
@@ -236,6 +236,14 @@ class Session {
 
     // Close cached image if it has been updated
     void CloseCachedImage(const std::string& directory, const std::string& file);
+    bool AnimationActive() {
+        return _animation_active;
+    }
+    void SetAnimationActive(bool val) {
+        fprintf(stderr, " SetAniActive %d\n", val);
```
Comment (veggiesaurus) [inline src/Session.h]: please use `spdlog::debug` for this

File: src/SessionManager/SessionManager.cc [hunk 3]
```diff
@@ -230,6 +229,7 @@ void SessionManager::OnMessage(WSType* ws, std::string_view sv_message, uWS::OpC
                 }
                 case CARTA::EventType::ANIMATION_FLOW_CONTROL: {
                     CARTA::AnimationFlowControl message;
+                    fprintf(stderr, " F ");
```
Comment (veggiesaurus) [inline src/SessionManager/SessionManager.cc]: please remove this

File: src/Concurrency.h [hunk 4]
```diff
@@ -83,21 +82,26 @@ class queuing_rw_mutex {
...
+        short wc2 = _writer_count;
```
Comment (veggiesaurus) [inline src/Concurrency.h]: What's the point of these two lines? (and the ones below). `rc` and `wc` are never used.

**Review by veggiesaurus — CHANGES_REQUESTED** [review]

File: src/Session.h (hunk 2 above)
Comment (confluence) [reply src/Session.h]: Please also use the full function name.

(Unless this was meant to be temporary and should just be removed.)

File: src/Session.cc [hunk 5]
```diff
@@ -2055,7 +2056,7 @@ bool Session::ExecuteAnimationFrame() {
 }
 
 void Session::StopAnimation(int file_id, const CARTA::AnimationFrame& frame) {
-    if (!_animation_object) {
+    if (_animation_object == nullptr) {
```
Comment (veggiesaurus) [inline src/Session.cc]: isn't this the same thing?

**Review by confluence — COMMENTED** [review]
I have the same comments and questions as @veggiesaurus ; otherwise this looks OK.

File: src/Concurrency.h (hunk 4 above)
Comment (robsimmonds) [reply src/Concurrency.h]: for debugging. gone now.

File: src/Concurrency.h [hunk 6]
```diff
@@ -90,14 +89,15 @@ class queuing_rw_mutex {
...
         if ((_reader_count == 0) && (_writer_count > 0)) {
-            --_writer_count;
             dequeue_one_writer();
         }
     }
     void writer_leave() {
         std::unique_lock<std::mutex> lock(_mtx);
-
         --_writer_count;
+        short rc = _reader_count;
```
Comment (veggiesaurus) [inline src/Concurrency.h]: these ones are still here?

**Review by veggiesaurus — CHANGES_REQUESTED** [review]

File: src/Concurrency.h (hunk 6 above)
Comment (robsimmonds) [reply src/Concurrency.h]: ok. i will remove and push. The tests have all completed successfully at this point.

**Review by confluence — APPROVED** [review]

**Review by veggiesaurus — APPROVED** [review]

**Pull request merged**